how-to-sans-io.rst

```diff
@@ -0,0 +1,328 @@
+~~~~~~~~~~~~~~~~~~
+
+When it comes to network protocols, at a fundamental level they all consume and
+produce byte sequences. For protocols implemented over TCP (or any
+``SOCK_STREAM``-type socket), they use a byte stream. For protocols implemented
+over UDP, or over any lower-level protocol than that (e.g. directly over IP),
+they communicate in terms of datagrams, rather than byte streams.
+
+For byte-stream based protocols, the protocol implementation can use a single
+input buffer and a single output buffer. For input (that is, receiving data
+from the network), the calling code is responsible for delivering code to the
+implementation via a single input (often via a method called ``receive_bytes``,
+or something similar). The implementation will then append these bytes to its
+internal byte buffer. At this point, it can choose to either eagerly process
+those bytes, or do so lazily at the behest of the calling code.
+
+When it comes to generating output, a byte-stream based protocol has two
+options. It can either write its bytes to an internal buffer and provide an API
+for extracting bytes from that buffer, as done by `hyper-h2`_, or it can return
+bytes directly when the calling code triggers events (more on this later), as
+done by `h11`_. The distinction between these two choices is not enormously
+important, as one can easily be transformed into the other, but using an
+internal byte buffer is recommended if it is possible that the act of receiving
+input bytes can cause output bytes to be produced: that is, if the protocol
+implementation sometimes automatically responds to the peer without user input.
+
+For datagram based protocols, it is usually important to preserve the datagram
+boundaries. For this reason, while the general structure of the above points
+remains the same, the inputs and outputs should be changed to consume and
+return iterables of bytestrings. Each element in the iterable will correspond
+to a single datagram.
+
+Events
+~~~~~~
+
+The major abstraction used by most of the sans-IO protocol stacks is to
+translate the bytes received from the network into "events". Essentially, this
+abstraction defines a network protocol as a serialization mechanism for a
+sequence of semantic "events" that can occur on that protocol.
+
+In this abstraction model, both peers in a protocol emit and receive events.
+In terms of receiving events, events can either be returned to the calling code
+immediately whenever bytes are provided, or they can be lazily produced in
+response to the calling code's request. Both approaches have their advantages
+and disadvantages, and it doesn't matter enormously which is chosen.
+
+When it comes to emitting events, there are several possible approaches, but
+two are in active use. The first, and comfortably the most common, is to emit
+events using function calls. For example, a HTTP implementation may have a
+function call entitled ``send_headers`` which emits a bytestream that, if
```

Comment:
Would it be worth adding the Sphinx extension that allows for referencing external docs so that these can directly link to hyper-h2 and h11?